server/credential_collection.cc

```diff
@@ -97,43 +67,36 @@ struct CollectionCredentialProvider: public CredentialProvider {
 
     }
 
-    std::shared_ptr<CredentialRuleCollection> rules;
+    // not owning because credential providers are stored in a global
+    // object and we don't want to delay the destruction of a collection
+    // object that late in the MLDB shutdown process
+    std::weak_ptr<CredentialRuleCollection> rules;
 
-    virtual std::vector<std::string>
-    getResourceTypePrefixes() const
+    virtual std::vector<StoredCredentials>
+    getCredentialsOfType(const std::string & resourceType) const
     {
-        auto keys =  rules->getKeys();
-        std::vector<std::string> result;
-        for (const auto & key : keys) {
-            auto cred = rules->tryGetEntry(key);
-            ExcAssert(!cred.second); // still under construction
-            if (cred.first)
-                result.push_back(cred.first->config->store->resourceType);
-        }
-        return result;
-    }
-
-    virtual std::vector<Credential>
-    getSync(const std::string & resourceType,
-            const std::string & resource,
-            const CredentialContext & context,
-            Json::Value extraData) const
-    {
-        vector<Credential> result;
-        auto keys =  rules->getKeys();
-        for (const auto & key : keys) {
-            auto cred = rules->tryGetEntry(key);
-            ExcAssert(!cred.second); // still under construction
-            if (cred.first) {
-                auto stored = cred.first->config->store;
-                if (resourceType != stored->resourceType)
-                    continue;
-                if (resource.find(stored->resource) != 0)
-                    continue;
-                result.push_back(stored->credential);
+        std::shared_ptr<CredentialRuleCollection> sharedRules =
+            rules.lock();
+
+        // getCredentialsOfType called after destruction of the collection!
+        ExcAssert(sharedRules);
+
+        vector<StoredCredentials> matchingCreds;
+        if (sharedRules) {
+            auto keys = sharedRules->getKeys();
+            for (const auto & key : keys) {
+                auto cred = sharedRules->tryGetEntry(key);
+                ExcAssert(!cred.second); // still under construction!
```

Comment:
It would be very unlikely since adding credentials is nothing like running a procedure.  It is probably why this was an assert instead of an http exception.